First cut at the revocation section

swcurran opened the pull request:
Signed-off-by: Stephen Curran <swcurran@gmail.com>

swcurran commented:
Does not include the cryptographic steps to generate or verify the proof, nor the data in the "non_revoc_proof" section of a presentation. All of that needs to be added, once this is merged.

wip-abramson commented on spec/data_flow_revocation.md:
````diff
@@ -1,11 +1,429 @@
+}
+```
+
+The item may at the outer level of the presentation request, such that it
````
```suggestion
The item may be at the outer level of the presentation request, such that it
```

wip-abramson replied:
sense check paragraph

swcurran replied:
Reviewed and clarified, with missing word added.

wip-abramson commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,429 @@
+  will use in generating the NRP. In the Hyperledger Indy implementation of
+  AnonCreds, the entries are published on the ledger, and collected via a
+  special request to the ledger (detailed below).
+- The revocation status changes of all of the credential indices up to the
```
delete `of`? Not sure. Sense check para

wip-abramson commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,429 @@
+are not received in the presentation, the [[ref: verifier]] SHOULD check to see
+if the given credential type is revocable. If not, it is acceptable that no
+NRP was received. However, if the credential used in the generation of the
+proof is revocable, and the [[ref: holder]] did not provide the P0N-R, the
```
```suggestion
proof is revocable, and the [[ref: holder]] did not provide the NRP, the
```

swcurran replied:
Fixed.

wip-abramson commented:
Reads well to me! Just made a few minor suggestions

swcurran commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,429 @@
+If the [[ref: issuer]] has created the [[ref: RevReg]] from which they are
+issuing credentials to have an `ISSUANCE_TYPE` of `ISSUE_ON_DEMAND`, the
+[[ref:issuer]] must publish a [[ref: RevRegEntry]] (as described in the next
+section) as each credential issued. Based on the experience of the AnonCreds
```
Instead changed to "as each credential is issued."

swcurran commented on spec/data_flow_revocation.md:
````diff
@@ -1,11 +1,429 @@
+]
+```
+
+As always, the [[ref: holder]]'s presentation, with the embedded NRP(s), is set to the
````
fixed.

swcurran commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,429 @@
+:::
+
+:::todo
+To Do: Is there a separate process to bind the NRP to the credential?
```
Not sure -- I think this needs to be an issue that is investigated. I'll add it.

swcurran commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,429 @@
+:::
+
+The verification code MUST surface to the [[ref: verifier]] if any part of the
+presentation, including any PoNR(s), fail cryptographic verification. The
```
fixed.

swcurran replied:
FYI - I'm making the fix in my copy vs. accepting the commit because I'm not sure how sign off is done. I don't want to screw up the DCO.  Someday I'll try it... :-)

swcurran commented:
Reworking section to move present proof components to the presentation section of the specification.

swcurran commented:
Updated per discussion at meeting on 2022.07.11. Migrated the sections related to issuing, presentation and verification into those parts of the specification. Retained the overall section to provide a summary of all parts of revocation in one place (with pointers to where the details are in other parts of the specification) and to cover the action of revoking (or activating) credentials.

Underlined the recommendation against using the "ISSUE_ON_DEMAND" option for issuing, where you have to update the RevReg whenever you issue a credential. Added a correlation danger for cases where "ISSUE_ON_DEMAND" is used and the credential has an "issue_date" type field in the credential. Verifiers knowing the issue_date can look on the ledger for a credential activated at that time and thus learn an identifier for the credential and monitor the revocation status of the credential.

@ArPhil -- please review -- particularly about the distribution of the revocation content into the other sections of the spec.

I'd like to get this merged so I can do some wholesale moving around of header levels in the specification -- move each of the issue, present, verify and revocation sections up one level.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+- An [[ref: issuer]] activating or revoking issued credentials.
+- A [[ref: verifier]] requesting a presentation to include a non-revocation proof
+  for one or more revocable credentials.
+- A [[ref: holder]] generating based on the request of the verifier a
```
```suggestion
- A [[ref: holder]] generating based on the presentation request of the verifier a
```

swcurran replied:
Fixed.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+  presentation from a [[ref: holder]].
+
+A fundamental goal of AnonCreds is to not provide a correlatable identifier for
+a either a [[ref: holder]] or a credential as part of generation and verification of an AnonCreds
```
```suggestion
either a [[ref: holder]] or a credential as part of generation and verification of an AnonCreds
```

swcurran replied:
Fixed.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+presentation. Applying that goal to revocation means that the revocation
+mechanism must support the [[ref: holder]] proving a credential used in generating a
+presentation is not revoked without providing a correlatable identifier for that
+credential. As such, the AnonCreds revocation mechanism uses a Zero Knowledge
```
```suggestion
credential or the holder itself. As such, the AnonCreds revocation mechanism uses a Zero Knowledge
```

swcurran replied:
Fixed.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+presentation is not revoked without providing a correlatable identifier for that
+credential. As such, the AnonCreds revocation mechanism uses a Zero Knowledge
+Proof (ZKP) that allows the [[ref: holder]] to prove a credential they hold is not
+revoked without revealing an identifier for their credential.
```
```suggestion
revoked without revealing an identifier for their credential or the holder.
```

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+holder]] are covered in the [issuance data
+flow](data_flow_issuance.md#issue-credential) section of this specification.
+
+If the [[ref: issuer]] has created the [[ref: RevReg]] from which they are
+issuing credentials to have an `ISSUANCE_TYPE` of `ISSUE_ON_DEMAND`, the
+[[ref:issuer]] must publish a [[ref: RevRegEntry]] (as described in the [next
+section](#anoncreds-credential-activationrevocation-and-publication)) as each
+credential is issued. Based on the experience of the AnonCreds community in the
+use of revocable credentials, it is highly recommended the `ISSUE_ON_DEMAND`
+approach **NOT** be used unless absolutely required by your use case. The reason
+for this recommendation is that `ISSUE_ON_DEMAND` requires near real-time
+activation, meaning the issuing of each credential requires the
+near-simultaneous publication of a new [[ref: RevRegEntry]], resulting in many
+[[ref: RevRegEntry]] transactions being performed, one per credential issued.
+
+Further, if a credential contains some kind of "Issue Date" attribute in the
+credential, verifiers could use that attribute information combined with the
+publication of a [[ref: RevRegEntry]] at the same time to learn the ID of the
+holder's credential, giving them both a correlatable identifier for the holder
+and a way to monitor if the holder's credential is revoked.
+
+For these reasons we anticipate the deprecation or removal of the
+`ISSUE_ON_DEMAND` approach in the next version of AnonCreds specification.
+Feedback from the community on this would be appreciated. We are particularly
+interested in understanding what use cases there are for `ISSUE_ON_DEMAND`.
```
```suggestion
::: warning
If the [[ref: issuer]] has created the [[ref: RevReg]] from which they are
issuing credentials to have an `ISSUANCE_TYPE` of `ISSUE_ON_DEMAND`, the
[[ref:issuer]] must publish a [[ref: RevRegEntry]] (as described in the [next
section](#anoncreds-credential-activationrevocation-and-publication)) as each
credential is issued. Based on the experience of the AnonCreds community in the
use of revocable credentials, it is highly recommended the `ISSUE_ON_DEMAND`
approach **NOT** be used unless absolutely required by your use case. The reason
for this recommendation is that `ISSUE_ON_DEMAND` requires near real-time
activation, meaning the issuing of each credential requires the
near-simultaneous publication of a new [[ref: RevRegEntry]], resulting in many
[[ref: RevRegEntry]] transactions being performed, one per credential issued.

Further, if a credential contains some kind of "Issue Date" attribute in the
credential, verifiers could use that attribute information combined with the
publication of a [[ref: RevRegEntry]] at the same time to learn the ID of the
holder's credential, giving them both a correlatable identifier for the holder
and a way to monitor if the holder's credential is revoked.

For these reasons we anticipate the deprecation or removal of the
`ISSUE_ON_DEMAND` approach in the next version of AnonCreds specification.
Feedback from the community on this would be appreciated. We are particularly
interested in understanding what use cases there are for `ISSUE_ON_DEMAND`.
:::
```

ArPhil replied:
Added a "warning" highlight box, since we do not recommend the usage of `ISSUE_ON_DEMAND`. I am also wondering, if we should not move this entire paragraph to the credential definition setup part with revocation capabilities. 

We than have to distinguish the optional revocation part within each section into a) `ISSUANCE_PER_DEFAULT` and b) `ISSUE_ON_DEMAND`. 

Or we only mention the two issuing modes once at the credential definition setup part and emphasize, that *ISSUE_ON_DEMAND' probably will not be supported in future versions and therefore don`t explain the details. 

Not sure yet what the best approach here is...

swcurran replied:
I like the idea of mentioning it with the warning in the setup section and have moved it there.  I think I will leave a warning and pointer to the primary warning.

swcurran replied:
Adjusted.  I moved the warning into the data setup, gave the warning a header, and then linked to that section in a couple of other places in the spec. I think that works pretty well.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+for a [[ref: RevReg]]. In that process, the accumulator for the initial state of
+the [[ref: RevReg]] is published. When subsequent [[ref: RevRegEntry]]
+transactions are published to the ledger, each includes an updated value of the
+accumulator, along with lists of the identifiers of the credentials that have
+had their revocation state toggled -- either changed to issued, or changed to
+revoked -- since the last [[ref: RevRegEntry]] was published. An example of the
+data in the [[ref: RevRegEntry]] is shown in the following example of a [[ref:
+RevRegEntry]], pulled from [this transaction on the Sovrin
+MainNet](https://indyscan.io/tx/SOVRIN_MAINNET/domain/140326).
```
```suggestion
accumulator. The update of the accumulator is necessary with each revocation or activation of a credential or set of credentials since the last published [[ref: RevRegEntry]]. This is because only the factors (all factors are listed in the respective tails file) of credentials which are active (meaning not being revoked) contribute to the accumulator. Therefore in addition to the updated accumulator value, every RevRegEntry contains a list of indexes of credential factors which have been revoked within the [[ref: RevRegEntry]]. This list of factor indexes is a so called [[ref: Witness Delta]] and enables the [[ref: Holder]] to successfully generate a proof of non revocation.

An example of the
data in the [[ref: RevRegEntry]] is shown in the following example of a [[ref:
RevRegEntry]], pulled from [this transaction on the Sovrin
MainNet](https://indyscan.io/tx/SOVRIN_MAINNET/domain/140326).
```

ArPhil replied:
I tried to differentiate between "identifiers of the credentials" and the "index of the factor within a tails file". From my point of view, this is crucial, since imho it is not some kind of ID of a credential, that is published within a RevRegEntry, but rather the index of the factor, that is bound to the credential. 

But I am not entirely happy with my wording here...

swcurran replied:
I consider the "RevocationID" to be the combined key `index+RevRegId`, and that the RevocationId is essentially a `CredentialId`.

**Question**: Does the `Witness Delta` include the factor contributed by the Credential that is the subject of the Non-Revocation Proof?  I've always wondered about that.  I think of it as the `Accumulator` = `Witness Delta` * `Credential Factor`

swcurran replied:
I liked your wording and put it in verbatim. I think we can merge with that and adjust if needed.

ArPhil replied:
> I consider the "RevocationID" to be the combined key `index+RevRegId`, and that the RevocationId is essentially a `CredentialId`.
> 
> **Question**: Does the `Witness Delta` include the factor contributed by the Credential that is the subject of the Non-Revocation Proof? I've always wondered about that. I think of it as the `Accumulator` = `Witness Delta` * `Credential Factor`

_This is handled by requiring accumulator updates to also publish a witness delta as part of the same transaction. This delta tells provers how to adjust their witness (referencing other indexes in the public tails file) to bring it back into harmony with the current value of the accumulator, such that the updated witness times the private factor once again equals the accumulator value._

(got it from [here](https://hyperledger-indy.readthedocs.io/projects/hipe/en/latest/text/0011-cred-revocation/README.html))

swcurran replied:
OK, so the witness is the product of all of the unrevoked factors excluding the one for the credential held by the holder. But the witness delta is exactly as you used it in your note above.  I've added the Witness Delta to the terminology section.

In looking at that reference, I also documented how the NRP ZKP is generated.  I hope it is enough detail, although I've noted that more detail is needed on generating the commitments.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+non-revoked credential in the [[ref: REV_REG]], as described
+[here](#creating-the-initial-revocation-registry-entry-object).
+
+A [[ref: VDR]] publishing the [[ref: RevRegEntry]] transactions MAY perform its own calculation
+of the accumulator to ensure that the calculation of the accumulator after all
+of the revocation status updates to the credentials within the [[ref: RevReg]] have been
+applied, rejecting the transaction if the calculated accumulator does not match that from the [[ref: issuer]].
```
Sounds wise to me, that the VDR does a double check on the updated accumulator value, but do we want to mention it here (in the AnonCreds Spec) or is it out of scope?

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+non-revoked credential in the [[ref: REV_REG]], as described
+[here](#creating-the-initial-revocation-registry-entry-object).
+
+A [[ref: VDR]] publishing the [[ref: RevRegEntry]] transactions MAY perform its own calculation
+of the accumulator to ensure that the calculation of the accumulator after all
+of the revocation status updates to the credentials within the [[ref: RevReg]] have been
+applied, rejecting the transaction if the calculated accumulator does not match that from the [[ref: issuer]].
+
+The [[ref: issuer]] MUST track of the revocation status of all of the credentials
+within a [[ref: RevReg]] so that it can both calculate the correct accumulator and send
+to the [[ref: VDR]] accurate lists (`issued` and `revoked`) of the indices of the
+credentials whose status has changed since the last [[ref: RevRegEntry]] was published.
+If the [[ref: issuer]] and [[ref: VDR]] get out of sync about the status of
+credentials in the [[ref: RevReg]], such that the two cannot agree on the current value
+of the accumulator, the [[ref:issuer]] must rationalize the differences and
+produce a [[ref: RevRegEntry]] transaction that accounts for both the last published
+[[ref: RevRegEntry]] published in the [[ref: VDR]] and the desired revocation status of all of the
+credentials in the [[ref: RevReg]].
+
+Note that the [[ref: holder]] is not involved in the credential revocation
+process. There is no technical requirement for an [[ref: issuer]] to notify the
+[[ref: holder]] that a credential they were issued has been revoked. That said, it is a
+courtesy that may improve the user experience of the [[ref: holder]]. [Aries RFC 0183
+Revocation
+Notification](https://github.com/hyperledger/aries-rfcs/tree/main/features/0183-revocation-notification)
+is an example of how that can be done. Even if not notified by the [[ref:
+issuer]] of the revocation of a credential, the [[ref: holder]] can detect their
+credential has been revoked when they retrieve the list of revoked credentials
+from the [[ref: VDR]] and discover the index of their credential in the list.
```
Sounds wise to me, that the VDR does a double check on the updated accumulator value, but do we want to mention it here (in the AnonCreds Spec) or is it out of scope?

I think the same about getting the VDR and issuer revocation list syncing and about informing the user about that his credential has been revoked. On the las one we are rather on the Aries level, than the Indy / AnonCreds level imho?

swcurran replied:
I think it is in scope as if the VDR (e.g. where ever it is published) and the issuer are not in sync, the holder will not be able to create a valid NRP. 

OK -- updated to reflect:

- If things get out of sync the NRP can't be created by the holder
- the VDR may prevent going out of sync
- the issuer must fix things if they do go out of sync.

I think that is reasonable to put into the spec.

ArPhil replied:
Sounds reasonable to me. I think I would add a `::: note` wrapper around it, just to indicate, that is some kind of "additional" info the reader gets here. Or maybe even a `::: warning` if we consider it as very important. 

To me the fact that the issuer has to keep things in sync is crucial because that is in the heart of AnonCreds. The info, that the VDR can assist not so much, because that depends on where (which VDR) AnonCreds are implemented in / to. How the issuer can fix when being out of sync (in Indy) sounds to me like an "problems / fixing" section. But maybe that is just me...

(Just trying to keep the AnonCreds part as much reduced as possible while adding "additional" separately) 😄

swcurran replied:
Good comments.  I've put a note wrapper on that as you suggest.

ArPhil commented on spec/data_flow_revocation.md:
```diff
@@ -1,11 +1,173 @@
+credential has been revoked when they retrieve the list of revoked credentials
+from the [[ref: VDR]] and discover the index of their credential in the list.
+
+#### AnonCreds Presentation Request with Revocation
+
+Creating an AnonCreds presentation is a two-step process, beginning with a
+request from the [[ref: verifier]] asking the [[ref: holder]] to include a
+non-revocation proof (NRP) in the presentation, and then the [[ref: holder]]
+creating the NRP and including it in the presentation sent to the [[ref:
+verifier]]. Both steps are covered here.
+
+Both the verifier requesting a non-revocation proof, and the [[ref: holder]]
+generating the non-revocation proof are covered in the sections of this
+specification about [requesting](#request-non-revocation-proofs) and
+[generating](#generate-non-revocation-proofs) presentations, respectively.
+
+#### AnonCreds Verification with Revocation
+
+A [[ref: verifier]] receives the presentation from the [[ref: holder]] and
+processes the [non-revocation-related parts of the presentation](#verify-presentation) and
+the [revocation-related parts of the presentation](#verify-non-revocation-proof)
+(if any) in the presentation. The resulting status of the presentation combines the
+verification outcomes from processing all proofs within the presentation. If 
+verification of one or more of the embedded proofs is unsuccessful, the
+presentation is rejected as unverifiable.
```
Personally, I would move these sections to the "Create Presentation Request" and "Verify Presentation" sections.

swcurran replied:
I moved the meat of the revocation sections over to the presentation and verification docs. I updated the wording to make it clear that this overview is only there to say -- "go look somewhere else".  I think that is reasonable.

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -1,15 +1,31 @@
+the [[ref: holder]] have an accompanying non-revocation proof (NRP). The generation of an NRP
+is described [in this section](#generate-non-revocation-proofs) of the specification.
+
+**NOTE**: Often in discussions about verifiable presentations, the term "[[ref: prover]]"
+is used to indicate the participant generating the request presentation.
+Throughout the Hyperledger Indy AnonCreds implementation the term `[[ref: prover]]` is
+used in the names of methods performed out by that participant. However, because
+in AnonCreds the [[ref: holder]] and the [[ref: prover]] are always the same entity, we'll use
+[[ref: holder]] to refer to the participant generating the requested presentation to
+emphasize that the same entity is both issued credentials and generating the
+presentations from those credentials.
+
```
```suggestion
::: note
Often in discussions about verifiable presentations, the term "[[ref: prover]]"
is used to indicate the participant generating the request presentation.
Throughout the Hyperledger Indy AnonCreds implementation the term `[[ref: prover]]` is
used in the names of methods performed out by that participant. However, because
in AnonCreds the [[ref: holder]] and the [[ref: prover]] are always the same entity, we'll use
[[ref: holder]] to refer to the participant generating the requested presentation to
emphasize that the same entity is both issued credentials and generating the
presentations from those credentials.
:::
```

swcurran replied:
Thanks! I keep forgetting about that.

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -353,8 +369,193 @@ The resulting presentation `proof_json` created by the Holder has the following
+Once the [[ref: holder]] has determined the required NRPs needed for the presentation,
+they must generate a NRP for each applicable source verifiable credential and
+add the NRPs to the presentation. For each, the [[ref: holder]] must collect the
+applicable [[ref: RevReg]] data published by the [[ref: issuer]] and then generate the proof.
```
```suggestion
applicable [[ref: RevReg]] respectively [[ref: RevRegEntry]]s data published by the [[ref: issuer]] and then generate the proof.
```

swcurran replied:
Adjusted the wording.

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -1,15 +1,31 @@
+is described [in this section](#generate-non-revocation-proofs) of the specification.
+
+**NOTE**: Often in discussions about verifiable presentations, the term "[[ref: prover]]"
+is used to indicate the participant generating the request presentation.
```
```suggestion
is used to indicate the participant generating the presentation request response.
```

swcurran replied:
Fixed.

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -353,8 +369,193 @@ The resulting presentation `proof_json` created by the Holder has the following
+  issued credential's [[ref: RevReg]]. The [[ref: holder]] likely (though not necessarily) would
+  have collected the tails file at the time of issuance. Recall (from this
+  section of the specification) that the tails file for a [[ref: RevReg]] is generated at
+  creation time and does not change.
```
```suggestion
  creation time and does never change.
```

swcurran replied:
Fixed and added link to tails section in the line above.

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -353,8 +369,193 @@ The resulting presentation `proof_json` created by the Holder has the following
+  have collected the tails file at the time of issuance. Recall (from this
+  section of the specification) that the tails file for a [[ref: RevReg]] is generated at
+  creation time and does not change.
+- The index of the credential within the [[ref: RevReg]] for the [[ref: holder]]'s specific
+  credential being used in the presentation. This information is given to the
+  [[ref: holder]] by the [[ref: issuer]] when the verifiable credential is issued.
```
Since the heading of this section says "Collecting RevRegEntry Data", I am not sure if the holders credential index should be listed here as well, because it also says that the section is about information which was published by the issure earlier.

Maybe we have either to change the heading to be more generic (e.g. "Required information for NRP generation") or place this peace on information somewhere else in order to avoid confusion?

swcurran replied:
I changed the title of the section -- good call!

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -353,8 +369,193 @@ The resulting presentation `proof_json` created by the Holder has the following
+  [[ref: holder]] by the [[ref: issuer]] when the verifiable credential is issued.
+- The accumulator published by the [[ref: issuer]] for the [[ref: RevRegEntry]] that the [[ref: holder]]
+  will use in generating the NRP. In the Hyperledger Indy implementation of
+  AnonCreds, the entries are published on the ledger, and collected via a
```
```suggestion
  AnonCreds, the entries are published as [[ref: RevRegEntry]]s on the ledger, and collected via a
```

swcurran replied:
Fixed.

ArPhil commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -353,8 +369,193 @@ The resulting presentation `proof_json` created by the Holder has the following
+- The timestamp (in the Unix epoch format) of the last [[ref: RevRegEntry]] in the
+  interval.
+
+Once collected, the [[ref: holder]] processes the `issued` and `revoked` lists
+to determine the credential status (revoked or not) of every credential in the
+[[ref: RevReg]]. As well, the [[ref: holder]] can at the point see if the
```
Do we have an example oh how these lists look like exactly? Just wondering, whether the list entries have some timestamp which the holder has to use for "ordering" the entries and figuring out, whether a credential (which maybe as been activated, revoked and activated again) is now active or revoked

swcurran replied:
The example in the data_flow_revocation.md file of a RevRegEntry includes the "lists".  Here is an example [RevRegEntry on Sovrin MainNet](https://indyscan.io/tx/SOVRIN_MAINNET/domain/140326).

ArPhil replied:
Thanks @swcurran !!

Looking at the `RevRegEntry` I see the transaction time besides the `revoked` and / or `issued` fields. But I also (somehwere read) that one can get all `RevRegEntry`s for a `RevReg` with only one request against the VDR. Not sure If I understood this correctly, but if so, the response has to contain either

a) an ordered list with all `RevRegEntries` (`revoked` and `issued` lists) so that the hilder / prover is able to figure out which credentials are / have been (re)activated and therefore knows which factors to consider for building the witness or

b) Only the (final) list of factors / indices of credential factors which contribute to the accumulator (the logic of figuring out, which factors have to be considered for building the current witness is done by the VDR in this case.

Maybe I have to look / read further for this...

swcurran replied:
Check out the [GET_REVOC_REG_DELTA](https://hyperledger-indy.readthedocs.io/projects/node/en/latest/requests.html#get-revoc-reg-delta) transaction. 

It's not clear how the lists are handled if an entry is both activated and revoked in the delta period with ordering the updates, but presumably the code is smart enough to ensure that from the lists, the holder can determine the correct revocation state of every credential in the registry.

ArPhil commented on spec/data_flow_presentation_create_request.md:
```diff
@@ -76,25 +76,58 @@ presents a decimal number (use `indy_generate_nonce` function to generate a corr
+as "Prove that your car insurance policy was not revoked on June 12, 2021 when
+the accident occurred."
+
+##### WQL Query
+
+The `restrictions` on each attribute/predicate defines a query filter in [[ref:
+Wallet Query Language]] (WQL) that is used to search the [[ref: holder]] wallet
+(storage) for candidate credentials that meet the restrictions. WQL is not
+required in an AnonCreds implementation, but the function it performs is
+required. That is, the [[ref: holder]] must scan the credentials it holds to
+find the ones that satisfy the requirements defined in the presentation request.
 
 The WQL specification can be found in
 [indy-sdk/docs/design/011-wallet-query-language/README.md](https://github.com/hyperledger/indy-sdk/blob/master/docs/design/011-wallet-query-language/README.md)
```
I somehow like, that the Wallet Query Language is mentioned and gives an idea how a holder (wallet) should be able to collect required credentials, but I am also not sure if we should mention it at all, because from point of view, that is outside the scope of AnonCreds. Just as you said: Some kind of mechanism is needed in order to be able to fetch required connections, but AnonCreds does not care (?) . :-)

swcurran replied:
I tend to agree, but I didn't write that part. I think it can be removed in a subsequent PR.  Further, even if we do keep it, it should be moved to the generating presentation section, not in the create_request section.

But we'll leave it for now.

ArPhil commented on spec/data_flow_presentation_verify.md:
```diff
@@ -49,3 +54,52 @@ pub extern fn indy_verifier_verify_proof(command_handle: CommandHandle,
+presentation failed, such as which NRP failed verification (if any).
+
+The [[ref: verifier]] SHOULD evaluate the presentation to make sure that the
+[[ref: holder]] provided all requested NRPs. Notably, if any expected PoNRs
```
```suggestion
[[ref: holder]] provided all requested NRPs. Notably, if any expected NRPs
```
changed for consistency

swcurran replied:
Fixed.

swcurran commented:
Good stuff-- thanks!  I'll go through these and flag when this is ready for a (hopefully) final review.

swcurran commented on spec/data_flow_presentation_create_presentation.md:
```diff
@@ -353,8 +369,193 @@ The resulting presentation `proof_json` created by the Holder has the following
+
+A `witness` is calculated in the same way as the accumulator (as described
+[here](#publishing-the-initial-initial-revocation-registry-entry-object)),
+except the revocation status of the credential being proven as not revoked is
```
Fixed.

swcurran commented on spec/data_flow_presentation_verify.md:
```diff
@@ -49,3 +54,52 @@ pub extern fn indy_verifier_verify_proof(command_handle: CommandHandle,
+
+If the presentation includes one or more Non-Revocation Proofs (NRPs) the
+[[ref: verifier]] must also extract from the verifiable presentation the NRPs
+and process each proofs. If any of the NRPs cannot be verified because one
```
Fixed.

swcurran commented:
Great comments!  I've pushed an update with all of these addressed.  Some of the issues I've resolved, since the update was trivial, others I've left if you want to see what I have done.

ArPhil commented:
@swcurran , I approved the PR, but my voice does not count in the AnonCreds github project, since I am not a reviewer with write access...  😅

swcurran commented:
Thanks -- at the next meeting, I'll propose that @ArPhil  be added as an Editor.

@TelegramSam @victormartinez-work -- could one of you review this PR and approve it?

I've added a couple of more tweaks based on @ArPhil 's comments.

TelegramSam reviewed (approved):
Discussed WG 20220725.
